src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -318,38 +180,38 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
         }
 
         // Internal and virtual for testing only
-        internal virtual async Task<VersionedSemanticRange> GetCSharpSemanticRangesAsync(
+        internal virtual async Task<IReadOnlyList<SemanticRange>?> GetCSharpSemanticRangesAsync(
             RazorCodeDocument codeDocument,
             TextDocumentIdentifier textDocumentIdentifier,
-            Range? range,
+            Range range,
             long documentVersion,
             CancellationToken cancellationToken,
             string? previousResultId = null)
         {
             var razorRanges = new List<SemanticRange>();
 
-            if (!TryGetMinimalCSharpRange(codeDocument, out var csharpRange))
+            if (!_documentMappingService.TryMapToProjectedDocumentRange(codeDocument, range, out var csharpRange))
             {
-                return new VersionedSemanticRange(razorRanges, IsFinalizedCSharp: true);
+                return new List<SemanticRange>(razorRanges);
             }
 
-            var csharpResponse = await GetMatchingCSharpResponseAsync(textDocumentIdentifier, documentVersion, csharpRange, cancellationToken);
+            var csharpData = await GetMatchingCSharpResponseAsync(textDocumentIdentifier, documentVersion, csharpRange, cancellationToken);
 
             // Indicates an issue with retrieving the C# response (e.g. no response or C# is out of sync with us).
-            // Unrecoverable, return null to indicate no change. It will retry in a bit.
-            if (csharpResponse is null)
+            // Unrecoverable, return empty list to indicate no change. It will retry in a bit.
+            if (csharpData is null)
             {
-                return VersionedSemanticRange.Default;
+                return null;
             }
 
             SemanticRange? previousSemanticRange = null;
-            for (var i = 0; i < csharpResponse.Data.Length; i += 5)
+            for (var i = 0; i < csharpData.Length; i += 5)
             {
-                var lineDelta = csharpResponse.Data[i];
-                var charDelta = csharpResponse.Data[i + 1];
-                var length = csharpResponse.Data[i + 2];
-                var tokenType = csharpResponse.Data[i + 3];
-                var tokenModifiers = csharpResponse.Data[i + 4];
+                var lineDelta = csharpData[i];
+                var charDelta = csharpData[i + 1];
+                var length = csharpData[i + 2];
+                var tokenType = csharpData[i + 3];
+                var tokenModifiers = csharpData[i + 4];
```

Comment:
Feel free to disagree if you don't like it, but for maintainability all of these `i + <x>` could just be `i++` (including the one in the for statement)

Reply:
In this scenario I actually like the explicit i + x because to Allison's point the semantic tokens are represented by 5 ints and this way allows any future debuggee to understand what index corresponds with what piece of data

Reply:
Happy to be overruled :)